src/storage/src/hummock/iterator/change_log.rs

```diff
@@ -0,0 +1,672 @@
+
+        Ok(())
+    }
+
+    async fn advance_to_find_oldest_epoch(&mut self) -> HummockResult<EpochWithGap> {
+        let mut ret = self.curr_key.epoch_with_gap;
+        debug_assert!(self.is_valid_epoch(ret));
+        self.new_value_iter.next().await?;
+        loop {
+            if !self.new_value_iter.is_valid() {
+                break;
+            }
+            let key = self.new_value_iter.key();
+            match self.curr_key.user_key.as_ref().cmp(&key.user_key) {
+                Ordering::Less => {
+                    // has advance to next key
+                    break;
+                }
+                Ordering::Equal => {
+                    assert!(ret > key.epoch_with_gap);
+                    if !self.is_valid_epoch(key.epoch_with_gap) {
+                        debug_assert!(self.min_epoch > key.epoch_with_gap.pure_epoch());
+                        break;
+                    }
+                    ret = key.epoch_with_gap;
+                    self.new_value_iter.next().await?;
+                    continue;
+                }
+                Ordering::Greater => {
+                    unreachable!(
+                        "hummock iterator advance to a prev key: {:?} {:?}",
+                        self.curr_key,
+                        self.new_value_iter.key()
+                    );
+                }
+            }
+        }
+        debug_assert!(self.is_valid_epoch(ret));
+
+        Ok(ret)
+    }
+
+    /// Advance the two iters to a valid position. After it returns with Ok,
+    /// it is possible that the position is valid but there is no change log value,
+    /// because the new and old value may consume each other, such as Insert in old epoch,
+    /// but then Delete in new epoch
+    async fn try_advance_to_next_valid(&mut self) -> HummockResult<()> {
+        // 1. advance the new_value_iter to the newest op between max and min epoch
+        self.advance_to_valid_key().await?;
+
+        if !self.is_current_pos_valid {
+            return Ok(());
+        }
+
+        // 2. advance new_value_iter to out of the valid range, and save the oldest value
+        let oldest_epoch = self.advance_to_find_oldest_epoch().await?;
+
+        // 3. iterate old value iter to the oldest epoch
+        self.is_old_value_set = false;
+        loop {
+            if !self.old_value_iter.is_valid() {
+                break;
+            }
+
+            let old_value_iter_key = self.old_value_iter.key();
+            match self
+                .curr_key
+                .user_key
+                .as_ref()
+                .cmp(&old_value_iter_key.user_key.as_ref())
+            {
+                Ordering::Less => {
+                    // old value iter has advanced over the current range
+                    break;
+                }
+                Ordering::Equal => match old_value_iter_key.epoch_with_gap.cmp(&oldest_epoch) {
+                    Ordering::Less => {
+                        assert!(
```

Comment:
Let me see whether I understand this assertion correctly. The reason why there shouldn't be old value between `oldest_epoch` and `min_epoch` is tnat:
- If the `oldest_epoch` represents an INSERT, we will hit the `Ordering::Less` branch in L262 because all epochs of the old value should be larger than the `oldest_epoch`.
- If the `oldest_epoch` represents a UPDATE/DELETE, we will hit the `Ordering::Equal` branch in L276 kbecause it is guaranteed to have an old value in the same epoch.

That being said, will we ever see an old value with the same user key but smaller epoch before breaking the loop? In other words, will L267 ever be hit?

Reply:
The iter_log may merge the operations of multiple epochs between the max and min epoch. 

Let's say we have epoch 1 and 2. In epoch1, we have an update with both new value and old value,  and and in epoch2 we have a delete with old value. In the iter_log of with epoch range as (epoch2, epoch2), the oldest epoch is epoch2. When we advance the old value iter, we will only see epoch1 first, which hits L276.

Reply:
Discussed offline: the assumption here is that for an old value in an epoch, there must be a new value with the same epoch. The assumption holds because old value is generated in UPDATE or DELETE and in either case, a new value will be written. Under this assumption, `old_value_iter_key.epoch_with_gap.pure_epoch() < self.min_epoch` must hold because otherwise, the `oldest_epoch` is not the "oldest".